Add "Forbidden" page for unsuccessful auth

jmhooper opened the pull request:
Fixes issue(s) # 213

Changes proposed in this pull request:

- Add error page that is displayed after unsuccessful sign in

This commit adds a page that is displayed when the user attempts to authenticate and fails.

The page provides some hints about where to look to determine what could be causing the problem.
Additionally it provides a link to attempt another sign in.

Originally I was going to render the new session view with a flash message, but decided on a separate page since it provides more information about what might have gone wrong.

jessieay commented:
Nice! Can we add a test for this?

jmhooper commented:
Definitely. Should this be covered by a controller spec, request spec, or both?

jessieay commented:
I'd be happy with either *or* a capybara feature spec

jmhooper commented:
@jessieay: Added some tests ✅✅

Weirdly enough it looks like one of the CI builds failed b/c of a spec that I didn't touch on this branch, but did on my other branch.

jessieay commented on app/views/auth/forbidden.html.erb:
```diff
@@ -0,0 +1,13 @@
+    <%= link_to "membership is public", "https://help.github.com/articles/publicizing-or-hiding-organization-membership/" %>.
+  <p/>
+  <p>
+    <%= link_to "Try to sign in with Github again", "/auth/githubteammember" %>
```
`GitHub` :)

jessieay commented on spec/support/oauth_helper.rb:
```diff
@@ -4,9 +4,9 @@ def login_with_oauth(user = create(:user))
   end
 
-  def setup_mock_auth(email = "test@example.com")
+  def setup_mock_auth(email = "test@example.com", overrides = {})
```
do we need overrides?

jmhooper replied:
I added them to test unsuccessful auth, but maybe I could change it to an `options` argument that handles something like `team_member?: false`?

jessieay commented on app/views/auth/forbidden.html.erb:
```diff
@@ -0,0 +1,13 @@
+    We were unable to authenticate your user profile.
+  </p>
+  <p>
+    If you are signing in via Github, please verify that you are a member of the appropriate Github organization and that your
```
`GitHub`

jessieay replied:
can we break these lines at 80 chars por favor?

jessieay commented on spec/features/create_user_spec.rb:
```diff
@@ -0,0 +1,27 @@
+require "rails_helper"
+
+feature "Create user" do
```
maybe we should make these more feature-y. So "User signs in" instead of "Create user" and "user is member of the correct GitHub team" rather than "creating a user..." (also perhaps we should rename file)

jessieay commented on spec/features/create_user_spec.rb:
```diff
@@ -0,0 +1,27 @@
+    visit root_path
+    click_link "Sign in with GitHub"
+
+    print page.html
```
debugger!

jessieay commented:
Oh yeah,I think that was another issue we discovered with code: CI fails in evening 😆 
Glad you opened that FIx!

jmhooper commented:
@jessieay: I changed `create_user_spec.rb` to `sign_in_user_spec.rb` and added some specs to cover signing in existing users.

I also removed the debug artifact.

jmhooper commented:
Okay, just pushed another fixup

- addressed the comment about the `overrides` hash by making it into an `options` hash
- addressed some of the issues in the auth/forbidden view

Also, TIL I've been spelling "GitHub" as "Github" my entire life without realizing it's supposed to be CamelCase 😳.

jessieay commented:
@jmhooper LOL CamelCase, sooooo oldschool GitHub

jessieay commented on spec/support/oauth_helper.rb:
```diff
@@ -4,11 +4,11 @@ def login_with_oauth(user = create(:user))
       OmniAuth::AuthHash.new(
         credentials: {
-          "team_member?" => true,
+          "team_member?" => options[:team_member].nil? ? true : options[:team_member],
```
we can change this in the future (no need to do it now), but I generally avoid these types of ternary statements. Reason: They are nice and compact but can be harder to understand.

jessieay commented:
Good one! this Looks great. Do you mind squashing and merging? That is not the process on every 18F project, but it's what we've been doing on this one: https://github.com/18F/dolores-landingham-slack-bot/blob/master/CONTRIBUTING.md#git-protocol

If you are not comfortable with squashing (can be scary) I am happy to walk you through it.
